### ISSUE TITLE
Notifier logs after upgrading to 2.17.4

### ISSUE THREAD

**Issue by beerlington** [sorentwo/oban#1040]
I'm upgrading my application from 2.16.3 -> 2.17.4 and am seeing this getting constantly logged out in development. I haven't tried it out on production because I wasn't sure if it's something to be concerned with or if it's a dev only thing.

```
[info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
[info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
```

What does this mean and how do I stop it?

The [upgrade guide](https://hexdocs.pm/oban/2.17.4/v2-17.html#content) is a little confusing too. It says

> "You must either [run the v12 migrations](https://hexdocs.pm/oban/2.17.4/v2-17.html#run-oban-migrations-for-v12-optional) or [disable insert triggers](https://hexdocs.pm/oban/2.17.4/v2-17.html#disable-insert-notifications-optional) in your Oban configuration, otherwise you'll receive duplicate insert notifications for each job."

In the docs, each method says it's optional, so is there one that is preferred over the other?

**Comment by sorentwo** [comment]
@beerlington Thanks for opening this issue. The constant warning was definitely a bug, as explained in the linked commit. We'll have v2.17.5 out soon with that fix.

> What does this mean and how do I stop it?

It is _supposed_ to mean that the notifier can't receive messages from other nodes in the cluster. It's only a concern in production. For now, you can safely ignore it (the mere fact that it was flipping means you had connectivity in the first place).

> In the docs, each method says it's optional, so is there one that is preferred over the other?

You should probably run the v12 migration to disable database triggers. It's up to you whether you want to keep any triggers at all (it's what allows execution immediately as jobs are inserted).

**Comment by lud-wj** [comment]
@sorentwo 

```
stale = System.system_time(:second) - state.interval * state.stale_mult
```

It seems that there is a subtraction of milliseconds to seconds.

**Comment by sorentwo** [comment]
@lud-wj You understand correctly, and I totally messed up that change. There's a fix in 6b9bf1d1176 along with a test for the logic. Thank you!

**Comment by noozo** [comment]
Although this is closed, i'm still getting those notifier log messages in my local environment after upgrading to the latest version (2.17.9). @sorentwo

**Comment by lud-wj** [comment]
@noozo If you see them only once it is expected.

**Comment by sorentwo** [comment]
Which notifier are you using? Those log entries are a sign that ping messages aren’t always being received, which indicates that something is wrong in the setup. Is it after a live reload, waking from sleep, or some other event?

**Comment by noozo** [comment]
my setup:
```
config :myapp, Oban,
  repo: MyApp.Repo,
  queues: MyAppObanQueues.queues(),
  plugins: [
    {Oban.Plugins.Cron, crontab: cron_jobs},
    # Prune jobs every hour
    {Oban.Plugins.Pruner, max_age: 3600}
  ]
```

queues() is just a list
cron_jobs is also a list of cron declarations.

**Comment by noozo** [comment]
everything else i think is just defaults. but actually now that i look at it running overnight, it seems they don't pop up recurring, though i get 2 messages per enqueued job. is that normal (albeit annoying)?

**Comment by apayu** [comment]
> i'm still getting those notifier log messages in my local environment after upgrading to the latest version (2.17.9)

 I'm encountering a similar situation when I start a iex on local, and I'm not sure if this is normal. (Oban latest version 2.17.9)
 
 ```
 iex -S mix phx.server
 ```

```[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"job staging switched to local mode. local mode polls for jobs for every queue; restore global mode with a functional notifier","mode":"local","source":"oban","event":"stager:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
[info] {"message":"job staging switched back to global mode","mode":"global","source":"oban","event":"stager:switch"}
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"job staging switched to local mode. local mode polls for jobs for every queue; restore global mode with a functional notifier","mode":"local","source":"oban","event":"stager:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
[info] {"message":"job staging switched back to global mode","mode":"global","source":"oban","event":"stager:switch"}
```

config

```
config :my_app, Oban,
  repo: MyApp.Repo,
  queues: [default: 10],
  plugins: [
    {Oban.Plugins.Pruner, max_age: 60 * 60 * 24 * 7}
  ]
```

```
postgres -V
postgres (PostgreSQL) 15.3 (Homebrew)
```

**Comment by sorentwo** [comment]
@apayu How close together were those messages? Are they spaced out every 15s?

**Comment by apayu** [comment]
This is the interval I observed in the logs.
```
# start
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
# about 1min
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
# about 30 sec
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
# about 30 sec
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
# about 1 min 45 sec
[info] {"message":"notifier can't receive messages from any nodes, functionality degraded","status":"isolated","source":"oban","event":"notifier:switch"}
[info] {"message":"job staging switched to local mode. local mode polls for jobs for every queue; restore global mode with a functional notifier","mode":"local","source":"oban","event":"stager:switch"}
# about 15 sec
[info] {"message":"notifier only receiving messages from its own node, functionality may be degraded","status":"solitary","source":"oban","event":"notifier:switch"}
[info] {"message":"job staging switched back to global mode","mode":"global","source":"oban","event":"stager:switch"}
```

**Comment by sorentwo** [comment]
@apayu The interval isn't consistent, so it doesn't seem like there's a bug in the detection (definitely not like what was originally happening in this issue).

**Comment by Munksgaard** [comment]
I am also experiencing these the log messages (on 2.17.10). Here's a sample:

```
18:52:20.191 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
18:54:20.211 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
18:54:20.212 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
18:59:35.259 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
18:59:35.259 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:01:05.274 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:01:05.274 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:02:20.287 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:02:20.500 [info] {"event":"stager:switch","message":"job staging switched to local mode. local mode polls for jobs for every queue; restore global mode with a functional notifier","mode":"local","source":"oban"}
19:02:35.290 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:02:35.540 [info] {"event":"stager:switch","message":"job staging switched back to global mode","mode":"global","source":"oban"}
19:03:05.296 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:03:05.296 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:04:50.315 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:04:50.315 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:05:20.321 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:05:20.321 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:05:50.327 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:05:50.327 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:09:20.361 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:09:20.361 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:10:05.368 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:10:05.368 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:10:35.374 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:10:35.374 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:11:05.379 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:11:05.575 [info] {"event":"stager:switch","message":"job staging switched to local mode. local mode polls for jobs for every queue; restore global mode with a functional notifier","mode":"local","source":"oban"}
19:11:50.388 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:11:50.697 [info] {"event":"stager:switch","message":"job staging switched back to global mode","mode":"global","source":"oban"}
19:12:20.394 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:12:20.395 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:14:05.412 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:14:05.413 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
19:18:42.142 [info] {"event":"notifier:switch","message":"notifier can't receive messages from any nodes, functionality degraded","source":"oban","status":"isolated"}
19:18:42.142 [info] {"event":"notifier:switch","message":"notifier only receiving messages from its own node, functionality may be degraded","source":"oban","status":"solitary"}
```

**Comment by sorentwo** [comment]
@Munksgaard Thanks for the extra logs. The timing info implies a race condition between the ping and the notification, which leads me to believe it's down to time warp/clock drift. [This commit](https://github.com/sorentwo/oban/commit/661dab94c858faad1f3a7f5a66ce4b1d56df071d) should solve the problem.

Feel free to give try by running from main in dev and let me know if it has any effect.

**Comment by jozuas** [comment]
I observed an issue equivalent to https://github.com/sorentwo/oban/issues/1040#issuecomment-2177810726 after upgrading to Oban 2.17.10. Pinning Oban in `mix.exs` to [this commit](https://github.com/sorentwo/oban/commit/661dab94c858faad1f3a7f5a66ce4b1d56df071d) fixed my problem.

**Comment by Munksgaard** [comment]
Thanks for your work @sorentwo. Might you consider making a new release?

**Comment by jozuas** [comment]
@Munksgaard was already done yesterday https://github.com/sorentwo/oban/releases/tag/v2.17.11

**Comment by Munksgaard** [comment]
Oh damn, sorry!
